[dynamo][autogradable leaf module] adding initial support

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #172395
* #172376
* #172372
* #172692
* #172152
* __->__ #170471


This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mlazos

ydwu4 committed:
[dynamo][autogradable leaf module] adding initial support

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



TODOS:
0. code cleaning up
1. add documentation
2. add more tests (logging, graph breaks, global variable, closures, pytree inputs, more modules, nested modules, unbacked input and output for fake_fn)
3. support decorating free functions (new pr)
4. support pytree outputs (new pr)
5. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

TODOS:
1.  code cleaning up
2. add documentation
3. add more tests (setattr, non-forward, logging, graph breaks, global variable mutations, closures, pytree inputs, more modules, nested modules, unbacked input and output for fake_fn)
4. support decorating free functions (new pr)
5. support pytree outputs (new pr)
6. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

TODOS:
1.  code cleaning up
2. add documentation
3. add more tests (setattr, non-forward, logging, graph breaks, global variable mutations, closures, pytree inputs, more modules, nested modules, unbacked input and output for fake_fn)
4. support decorating free functions (new pr)
5. support pytree outputs (new pr)
6. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

TODOS:
1.  code cleaning up
2. add documentation
3. add more tests (setattr, non-forward, logging, graph breaks, global variable mutations, closures, pytree inputs, more modules, nested modules, unbacked input and output for fake_fn)
4. support decorating free functions (new pr)
5. support pytree outputs (new pr)
6. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

TODOS:
1.  code cleaning up
2. add documentation
3. support decorating free functions (new pr)
4. support pytree outputs (new pr)
5. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

Follow-up PRs:
1. support decorating free functions (new pr)
2. support pytree outputs (new pr)
3. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

Follow-up PRs:
1. support decorating free functions (new pr)
2. support pytree outputs (new pr)
3. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

Follow-up PRs:
1. support decorating free functions (new pr)
2. support pytree outputs (new pr)
3. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

Follow-up PRs:
1. support decorating free functions (new pr)
2. support pytree outputs (new pr)
3. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/config.py:
```diff
@@ -39,6 +39,11 @@
+# Validate that fake_fn and real_fn in @leaf_function decorators produce outputs
+# with matching shapes and dtypes in eager mode. Helps catch mismatches early.
+# [@compile_ignored: debug]
+validate_leaf_function_outputs = False
```
Flipped it to True. It has some overhead when turning it on (i.e. we need to run both real and fake impl at runtime). Added that into the docstring.

ydwu4 commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +239,127 @@ def deregister() -> None:
+          the same pytree structure, tensor shapes, and dtypes as real_impl. Enable
+          ``torch._dynamo.config.validate_leaf_function_outputs = True`` to validate.
+        - (Temporary) Only nn.Module methods supported: Currently designed for methods
+          on ``torch.nn.Module``, not standalone functions.
```
We'll add it as a follow up pr! It takes a different path than module method in dynamo.

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +239,152 @@ def deregister() -> None:
+        trace_rules._allowed_callable_ids.remove(wrapped_id)
+        trace_rules._leaf_function_ids.remove(wrapped_id)
+
+    weakref.finalize(inner, deregister)
```
Can someone remind me why we use ids and not weakrefs to maintain these sets?

ezyang replied:
Well I guess if it's symmetric with the other sites...

ezyang commented on torch/_dynamo/trace_rules.py:
```diff
@@ -3784,8 +3795,29 @@ def f3(x, y):
+
+
+@contextlib.contextmanager
+def force_inline() -> Iterator[None]:
```
bro you need some docs

ezyang commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1429,6 +1429,11 @@ def call_function(
 
+        if is_leaf_function(self.fn):
+            call_args = [*self.self_args(), *args]
+            var = variables.TorchInGraphFunctionVariable(self.fn, leaf_function=True)
```
It's kind of interesting that TorchInGraphFunctionVariable can be reused in this way, not obvious to me!

ydwu4 replied:
lol, yeah, this follows how nonstrict_trace is implemented.

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,481 @@
+    Context manager to temporarily allow non-fake inputs in fake tensor mode.
+
+    This is needed for leaf functions where fake_impl may capture real constant
+    tensors in its closure. These will be auto-converted to fake tensors.
```
This seems... bad? Why do we not force people to not reference real tensors in their fake impl?

ezyang replied:
If the fake impl closes over a real tensor that is not global you are leaking it!

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,481 @@
+    version_before: list[int],
+) -> None:
+    """
+    Check that no input tensors were mutated in-place during leaf function execution.
```
Can I mutate buffers inside the leaf module region?

ydwu4 replied:
no.. It will throw an error in the input mutation check. the leaf_module's named buffers have been flattened and is part of the inputs so the input mutation checks apply to the buffers.

ydwu4 replied:
just to confirm, is the requirement that we would like to 1. tighten _check_input_mutation to ban inputs but allow buffer mutations and 2. give user a config to disable input mutation checks?

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,481 @@
+def _validate_outputs_match(
+    fake_output: Any,
+    real_output: Any,
+) -> None:
```
I'm not reviewing this

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,481 @@
+
+@invoke_leaf_function.py_impl(ProxyTorchDispatchMode)
+def invoke_leaf_function_proxy_mode(proxy_mode, *all_args):
+    out = invoke_leaf_function(*all_args)
```
What will this typically dispatch to here?

ydwu4 replied:
usually dispatch to its fake implementation.

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,481 @@
+        with torch._C._ForceDispatchKeyGuard(effective_keys, exclude_keys):
+            with torch.enable_grad():
+                state["outputs"] = invoke_with_flattened_inputs(fn, input_spec, *args)
+        return state["outputs"]
```
It seems undesirable to save the output tensors directly, as they may not actually be needed to save for backwards. Instead, we should save only the graph EDGE, which is sufficient to trigger torch.autograd.grad. cc @albanD @soulitzer

ezyang replied:
Similarly, it seems the inputs are saved as well

ezyang commented:
I'm not qualified to review the Dynamo support but the HOP here mostly looks like it's in order. I am not approving to ensure Dynamo folks have the ability to review.

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

Follow-up PRs:
1. support decorating free functions (new pr)
2. support pytree outputs (new pr)
3. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/trace_rules.py:
```diff
@@ -3784,8 +3795,29 @@ def f3(x, y):
+    if _force_inline:
+        return SkipResult(
+            False,
+            "don't skip since it's marked as force inline",
```
yeah, updated the reason

ydwu4 commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,481 @@
+
+    nn_module_index: int
+    named_parameters: dict[str, torch.nn.Parameter]
+    named_buffers: dict[str, torch.Tensor]
```
This is actually just a pytree container that helps us track which parameters/buffers belong to which module. The params/buffers themselves get flattened as graph inputs - this structure just lets us reconstruct the right module at runtime. So no memory being held here.

ydwu4 commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,481 @@
+    Reconstruct original (args, kwargs) from flattened arguments.
+
+    Handles LeafModuleState by retrieving the original module and reparametrizing
+    it with the parameters/buffers arguments.
```
Good catch. Haven't tested with DTensor. I can do it in next pr. This pr is a bit too long already lol.

ydwu4 commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,481 @@
+    structure (e.g. nn modules are restored).
+    """
+    with reconstruct_original_args(input_spec, flat_args) as (args, kwargs):
+        return fn(*args, **kwargs)
```
Fair, inlined the helper function.

Lucaskabela commented on torch/_dynamo/variables/torch.py:
```diff
@@ -2232,6 +2248,141 @@ def flat_apply_capture(*args: Any) -> list[object]:
+        args: Sequence[VariableTracker],
+        kwargs: "dict[str, VariableTracker]",
+    ) -> tuple[
+        "VariableTracker",
```
Similar comment - no need to wrap these in quotes

Lucaskabela commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +239,151 @@ def deregister() -> None:
     return wrapped
 
 
+def leaf_function(fn: Callable[..., Any]) -> Callable[..., Any]:
```
fn should be typed with `P` as ParamSpec and `R` for return value to ensure we don't erase the types in fn - search for `: P` or `ParamSpec` to see how this pattern is often used

Lucaskabela commented on torch/_dynamo/variables/torch.py:
```diff
@@ -2232,6 +2248,141 @@ def flat_apply_capture(*args: Any) -> list[object]:
+        from .higher_order_ops import _make_inlined
+        from .nn_module import NNModuleVariable, UnspecializedNNModuleVariable
+
+        def is_module_variable(var: VariableTracker) -> bool:
```
This can be a `TypeIs` annotation (can also search for these)

Lucaskabela reviewed (changes requested):
Core logic looks good but had some comments on the typing as well as defering the runtime error to a later TODO

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

Follow-up PRs:
1. support decorating free functions (new pr)
2. support pytree outputs (new pr)
3. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -507,17 +507,28 @@ def __init__(
             nonstrict_traceable = is_nonstrict_trace_callable(value)
+        if leaf_function is None:
+            leaf_function = is_leaf_function(value)
+            # TODO: raise a proper error if a function is
```
added the check in the annotator, so we could check the conflict at module initialization time.

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a method of an nn Module as leaf function, dynamo and aot won't trace into the function. This PR only supports tuple outputs. 

Follow-up PRs:
1. support decorating free functions (new pr)
2. support pytree outputs (new pr)
3. support input mutation/aliasing support (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

Lucaskabela reviewed (approved):
Thanks for incorporating feedback - I think the logic looks good and changes look clean, approving!

williamwen42 reviewed (approved):
dynamo changes lgtm

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

albanD commented:
A lot of the LOC is from expecttest results which are auto-generated. So if @ezyang is happy with the PR, it's ok to skip the sanity check.

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -217,6 +217,13 @@ def nonstrict_trace(traceable_fn: Callable[_P, _R]) -> Callable[_P, _R]:
+        raise ValueError(
+            f"Function {traceable_fn} cannot be both marked as @leaf_function and "
+            f"@nonstrict_trace. Please use only one decorator."
+        )
```
One off testing this here feels unscalable. What happens when I apply the decorators in the other order?

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+    compilation with fake tensors (which have shapes/dtypes but no actual values)
+    and must return outputs with the same pytree structure, shapes, and dtypes as
+    the real function. Note: fake_impl may run multiple times during compilation;
+    real_impl runs once per call at runtime.
```
It's worth mentioning that, BECAUSE the fake impl must be able to run ahead of time, that means the out signature of the leaf module must be known at compile time. If you have some code where it's not known what the output shape/spec would be, you should just expand the scope of the leaf region until it is.

I am also interested to know if unbacked symints work on output.

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+    Can be used in two ways:
+    1. @leaf_function - uses the decorated function itself as fake_impl. This works
+       when the function can run with fake tensors (avoids data-dependent control
+       flow, etc.), avoids dangerous patterns, and meets the Restrictions below.
```
This mode of use is confusing to me. In particular, if you are able to directly propagate fake tensors through the implementation inside without problems, it seems like to me that an AOT trace *would* work. So it would seem better to make this NOT the default mode of use and force people to write fake impls (and point them towards `allow_in_graph` if the computation graph is static?)

ydwu4 replied:
sg, make fake_impl a requirement.  added a section that helps users to choose between leaf_function and allow_in_graph.

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+
+    Note:
+        - Training is supported e.g. you can call ``.backward()`` on outputs and gradients
+        will flow through the leaf function.
```
But it's not supported to call backward inside of the leaf function! lol

ydwu4 replied:
you're right! Added it as well. Also added a how it works section for the detach input and output design so that users could have a better mental model.

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +239,127 @@ def deregister() -> None:
+              @leaf_function
+              def forward(self, x):
+                  x.add_(1)  # In-place mutation are dangerous
+                  return (x,)
```
OK so update the docs here now?

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+
+              @leaf_function
+              def my_leaf_fn(x):
+                  x.add_(1)  # In-place mutation on input is dangerous
```
Don't say "dangerous", be specific about what will go wrong

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+          same primitives and structure on every execution.
+        - (Temporary) Return value should be a tuple of tensors.
+
+    Example:
```
I'd greatly appreciate it if the examples showed *realistic* uses of the API, e.g., when you would actually do this as opposed to using another API which would be more appropriate. For example...

ydwu4 replied:
Fair. Added some logging, external_lib examples!

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+                self.linear = torch.nn.Linear(10, 10)
+
+            def forward(self, x):
+                return my_leaf_fn(self.linear, x)
```
Just trace this directly!

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+        def train_step_fn(model, x, target):
+            pred = model(x)
+            loss = torch.nn.functional.mse_loss(pred, target)
+            return (loss,)
```
Just Dynamo trace this!

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,489 @@
+        fake_tensor_tls.allow_non_fake_inputs_override = True
+        yield
+    finally:
+        fake_tensor_tls.allow_non_fake_inputs_override = old_override
```
I am really unsure why this was so important as to have needed to be put into this first version

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,489 @@
+    @functools.wraps(fn)
+    def forward(*args):
+        state["inputs"] = tuple(
+            arg.requires_grad_(True) if idx in requires_grad_indices else arg
```
Mutating arg directly here seems dangerous. Why do you know it is safe to do this?

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,489 @@
+        if include_keys.has(DispatchKey.PythonDispatcher):
+            current_include = torch._C._dispatch_tls_local_include_set()
+            if not current_include.has(DispatchKey.PythonDispatcher):
+                effective_keys = include_keys.remove(DispatchKey.PythonDispatcher)
```
I don't understand what this is doing, but it doesn't feel good

ezyang commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,489 @@
+) -> tuple[Callable, Callable]:
+    """
+    Create forward/backward wrappers for tracing.
+    """
```
Why do we need both this and `make_runtime_wrappers`?

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+        Inductor backend and ``torch.export`` are not yet supported.
+
+    Dangerous patterns (apply to both usage options, may cause silent incorrectness):
+        - Side effects: Don't rely on side effects for correctness.
```
sg! added a how it works section.

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ezyang commented on torch/_dynamo/decorators.py:
```diff
@@ -239,6 +246,213 @@ def deregister() -> None:
+          For inference-only use cases where gradients are not needed, you can set
+          ``torch._dynamo.config.leaf_function_allow_non_fake_inputs = True`` to allow
+          constant tensor closures. However, gradients will NOT flow to captured tensors/modules,
+          so do not use this for training.
```
This situation seems detectable. You're specifically looking to see if the known inputs dominate the output `grad_fn` graph. If you're able to escape, that means that there's some other things that require grad that escape, and that is bad. Very similar to the discussion with @tugsbayasgalan in https://github.com/pytorch/pytorch/pull/169493

Interestingly, if the outer autograd was a backward(), you could actually just trigger a backward() here ala reentrant checkpointing and this would work, but I think people generally decided this was too complicated for people to understand. @albanD probably isn't too interested in adding an option for the autograd to discover NEW grad inputs as the graph is executing to the engine LOL...

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
         with self.assertRaises(RuntimeError):
             exported_model = torch.export.export(model, (inp,))
 
+    def _assert_models_equal(
```
@claude Look over all the new tests in this file and summarize on a test-by-test basis what they test.

ezyang commented:
@claude Describe the high level control flow when I execute an InvokeLeafFunction HOP. Control should be passing from function to function in this file. Describe how it all flows together.

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+
+        real_fn : torch.utils._pytree.TreeSpec = self.real_fn
+        fake_fn : torch.utils._pytree.TreeSpec = self.fake_fn
+        non_tracable_forward_input_spec : torch.utils._pytree.TreeSpec = self.non_tracable_forward_input_spec
```
@angelayi since when have we been able to put TreeSpec in FX graphs! lol

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+        real_fn : torch.utils._pytree.TreeSpec = self.real_fn
+        fake_fn : torch.utils._pytree.TreeSpec = self.fake_fn
+        non_tracable_forward_input_spec : torch.utils._pytree.TreeSpec = self.non_tracable_forward_input_spec
+        invoke_leaf_function = torch.ops.higher_order.invoke_leaf_function(real_fn, fake_fn, non_tracable_forward_input_spec, 0, l_self_modules_linear_parameters_weight_, l_self_modules_linear_parameters_bias_, l_x_);  real_fn = fake_fn = non_tracable_forward_input_spec = l_self_modules_linear_parameters_weight_ = l_self_modules_linear_parameters_bias_ = l_x_ = None
```
What does the 0 mean here?

ydwu4 replied:
It's the nn_module_index of flattened LeafModuleState

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+    def forward(self, tangents_1: "f32[3, 3]"):
+        _tree_spec_constant2 = self._tree_spec_constant2
+        _tree_spec_constant3 = self._tree_spec_constant3
+        invoke_leaf_function_1 = torch.ops.higher_order.invoke_leaf_function(_tree_spec_constant2, _tree_spec_constant3, None, tangents_1);  _tree_spec_constant2 = _tree_spec_constant3 = tangents_1 = None
```
The lack of data dependency here is unexpected. I'm expecting leaf functions to generate an opaque object representing the autograd graph that we froze in time, which then gets transported to the backwards to be run.

I am kind of wondering whether or not any of the tests here test that I can run fwd/bwd of a graph leaf function *twice*.

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+""",  # noqa: B950
+        )
+
+    def test_leaf_function_in_inner_module(self):
```
This test is kind of pointless, why do I expect being a leaf function in an inner module is going to be any different from... well, anywhere else? (This feels like a holdover from the older version of this PR where you directly wrapped NN moduels)

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+        from torch._dynamo.decorators import leaf_function
+
+        @leaf_function
+        def custom_method(mod, x):
```
It's not a method though no?

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+            self._test_leaf_function_helper(LoggingModule, args_fn, loss_fn)
+            mock_print.assert_any_call("Processing input")
+            # Called 3 times: eager, compile_eager, and compile_aot
+            self.assertEqual(mock_print.call_count, 3)
```
This raises a good question. If a leaf function has no outputs, will it get DCE'd? I don't see anything that obviously prevents DCE when the outputs don't end up getting used.

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+            # Called 3 times: eager, compile_eager, and compile_aot
+            self.assertEqual(mock_print.call_count, 3)
+
+    def test_leaf_function_global_and_closure_read_only(self):
```
"read only" doesn't express that the globals in question do NOT require grad

ydwu4 replied:
Fair! Changed the wording.

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+
+        @outer_leaf_forward.fake_impl
+        def outer_leaf_forward_fake(mod, x):
+            return mod.inner(mod.linear(x))
```
I know I did say I was OK with only doing the functional version first, but it sure is weird seeing EVERY SINGLE TEST dealing with modules... like... surely you can do some tests that don't involve modules??

ydwu4 replied:
lol, we have one test for non-module input. Added one more with pytree inputs without nn module.

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+        )
+
+    def test_leaf_function_sequential_module_list(self):
+        """Test ModuleList where each module calls a leaf_function."""
```
I also doubt this variation (a ModuleList specifically) actually makes any difference

ezyang commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+            normalize_gm(eager_backend.graphs[0].print_readable(print_output=False)),
+            normalize_gm(backend.fw_graphs[0].print_readable(print_output=False)),
+            normalize_gm(backend.bw_graphs[0].print_readable(print_output=False)),
+        )
```
The testing here seems a bit naive. In particular, I would guess these tests would all pass even if you replaced `leaf_module` with `allow_in_graph`.

You need to verify that we are actually building the autograd graph *every* single time you run the graph, and haven't just burned in one particular graph that you saw when you traced or whatever. To do that, you need to run the program multiple times, and you need a leaf function that will actually do something different in autograd each iteration, so you can actually observe that the compiled graph is doing the right thing on each iteration.

I don't think you need a fancy generalized helper, one or two tests would be good enough for confidence here IMO.

ydwu4 replied:
Added two tests test_leaf_function_dynamic_autograd_module_config and test_leaf_function_dynamic_autograd_closure that changes the model config/closure to affect the behavior of leaf_function.

ydwu4 commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+
+        mod_eager = mod_class()
+        mod_compile_eager = mod_class()
+        mod_compile_eager.load_state_dict(dict(mod_eager.state_dict()))
```
mod_compile_eager will copy_() the paramters and buffers of mod_eager. So params and buffers have the same values but are not aliasing.

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,489 @@
+
+
+def _retrieve_module_by_index(nn_module_index: int) -> torch.nn.Module:
+    from torch._dynamo.graph_bytecode_inputs import get_external_object_by_index
```
Revert the import by making dynamo set a retriever callback for leaf_function to call.

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token
4. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 commented on test/dynamo/test_decorators.py:
```diff
@@ -2214,6 +2219,1355 @@ def forward(self, x) -> torch.Tensor:
+        @nonzero_forward.fake_impl
+        def nonzero_forward_fake(mod, x):
+            out = mod.linear(x)
+            return out, (out > 0).nonzero()
```
This doesn't have to be happen in this PR, right?

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

zou3519 reviewed (commented):
I think the main two things I didn't understand were:
1) what is the autograd strategy
2) how we are pulling the modules from their source to be called later on

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

zou3519 commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,564 @@
+    def __init__(self):
+        super().__init__("invoke_leaf_function")
+
+    def __call__(self, real_fn_spec, fake_fn_spec, input_spec, *flat_args):
```
docstring please. it should clarify that the fake_fn is the function registered via register_fake

zou3519 reviewed (commented):
I think I have a suggestion to make the autograd stuff better, let me know if that makes sense. Otherwise this lgtm

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 commented on torch/_higher_order_ops/invoke_leaf_function.py:
```diff
@@ -0,0 +1,564 @@
+    flat_args = _detach_tensors(flat_args)
+    with reconstruct_original_args(input_spec, flat_args) as (args, kwargs):
+        out = fn(*args, **kwargs)
+    return _detach_tensors(out)
```
Yeah, you're right! We should do the detach within the _make_forward.

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 committed:
Update on "[dynamo][autogradable leaf module] adding initial support"



This pr supports annotating a function as leaf function, dynamo and aot won't trace into the function. Nn module inputs are allowed. This PR only supports tuple outputs. 

Follow-up PRs:
1. support pytree outputs (new pr)
2. support dtensor (new pr)
3. support effect token (new pr)
4. support buffer mutation within leaf_function (new pr)
5. support input mutation/aliasing (new prs)
6. get a fake context like custom ops.


cc albanD voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx kadeng chauhang amjames Lucaskabela jataylo mlazos

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge
